object_intern.go

```diff
@@ -168,44 +212,104 @@ func (oi *ObjectIntern) AddOrGet(obj []byte, safe bool) (uintptr, error) {
 // If the object is found in the store its reference count is increased by 1.
 // If the object is added to the store its reference count is set to 1.
 func (oi *ObjectIntern) AddOrGetString(obj []byte, safe bool) (string, error) {
-	var addr gos.ObjAddr
-	var ok bool
-	var err error
 
-	objComp := obj
+	// if either of these two terms is true then the rest of this block
+	// requires a lot of allocations
+	if (oi.conf.Compression != None) || (safe && oi.conf.Compression == None) {
+		objComp := obj
 
-	if oi.conf.Compression != None {
-		objComp = oi.compress(obj)
-	}
+		if oi.conf.Compression != None {
+			objComp = oi.compress(obj)
+		}
 
-	// the only case we need to handle specially is when compression
-	// is turned off and the user has requested that the operation be safe
-	if safe && oi.conf.Compression == None {
-		// create a copy so we don't modify the original []byte
-		objComp = make([]byte, len(obj))
-		copy(objComp, obj)
-	}
+		// the only case we need to handle specially is when compression
+		// is turned off and the user has requested that the operation be safe
+		if safe && oi.conf.Compression == None {
+			// create a copy so we don't modify the original []byte
+			objComp = make([]byte, len(obj))
+			copy(objComp, obj)
+		}
+
+		// acquire lock
+		oi.Lock()
```

Comment:
Need to take a closer look at where we are locking and unlocking, might be able to reduce lock times